File: 147. Insertion Sort List/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertionSortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    curr1 := head.Next
    prev1 := head
    for curr1 != nil {
        curr2 := head
        var prev2 *ListNode
        
        // Identify the sorted position for this element (runner)
        for curr2 != curr1 {
            if curr2.Val > curr1.Val {
                break
            } else {
                prev2 = curr2
                curr2 = curr2.Next
            }
        }
        
        if curr2 == curr1 {
            // The element is already in correct position
            prev1 = curr1
            curr1 = curr1.Next
        } else {
            // move the element pointed to by runner to the above position
            tmp := curr1.Next
            prev1.Next = tmp
            
            if prev2 == nil {
                curr1.Next = head
                head = curr1
            } else {
                prev2.Next = curr1
                curr1.Next = curr2
            }
            
            curr1 = tmp
        }
    }
    return head
}

File: 147. Insertion Sort List/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def insertionSortList(self, head: 'ListNode') -> 'ListNode':
        if not head:
            return head
        
        # create a dummy head and tail
        tmp = head
        head = ListNode(-sys.maxsize-1)
        tail = ListNode(sys.maxsize)
        head.next = tail
        tail.next = tmp
        
        while tail.next:
            # extract node
            tmp = tail.next
            tail.next = tmp.next
            tmp.next = None
            
            # insert node to right position
            node = head
            while node.next.val < tmp.val:
                node = node.next
                
            tmp.next = node.next
            node.next = tmp
            
        # remove dummy tail
        node = head
        while node.next.next:
            node = node.next
        node.next = None
        
        # remove dummy head
        return head.next
            
